File: C_Sharp/Treap/Treap.cs
using System;
using Treap.Interfaces;

namespace Treap
{
    [Serializable]
    public class Treap<T> : ITreap<T> 
    {
        private int count;
		private BaseTreapNode<T> treapTree;

        public void Insert(int idx, T data)
		{
            if (idx < 0 || idx > count)
            {
                throw new ArgumentException(string.Format("Out of range. Idx = {0}, count = {1}", idx, count), "idx");
            }

		    InsertNodeInternal(idx, data);
		}

        internal BaseTreapNode<T> InsertNodeInternal(int idx, T data)
        {
            BaseTreapNode<T> node = CreateNode(data);
            treapTree = InsertNode(idx, node, treapTree);
            count++;
            return node;
        }


        protected virtual BaseTreapNode<T> CreateNode(T data)
        {
            return new BaseTreapNode<T>(data);
        }

        protected BaseTreapNode<T> TreapTree
        {
            get { return treapTree; }
        }

        public T this[int idx]
	    {
	        get
	        {
                ValidateIdx(idx);
                
	            BaseTreapNode<T> treeNode = treapTree;

	            while (treeNode != null)
	            {
	                int leftTreeCount = treeNode.Left != null ? treeNode.Left.Count : 0;

	                if (idx == leftTreeCount)
	                {
	                    return treeNode.Value;
	                }
	                if (idx < leftTreeCount)
	                {
	                    treeNode = treeNode.Left;
	                }
	                else
	                {
	                    idx -= leftTreeCount + 1;
	                    treeNode = treeNode.Right;
	                }
	            }

	            throw (new ArgumentException("Treap key was not found"));
	        }

	        set
	        {
	            ValidateIdx(idx);
                Set(treapTree, idx, value);
	        }
	    }

        public void Move(int source, int dest)
        {
            T temp = this[source];
            Delete(source);
            Insert(dest, temp);
        }

        public int Count
        {
            get { return count; }
        }


        public void Delete(int idx)
        {
            ValidateIdx(idx);

            treapTree = Delete(idx, treapTree);
            count--;
        }

        protected void ValidateIdx(int idx)
        {
            if (idx < 0 || idx >= count)
            {
                throw new ArgumentException(string.Format("Out of range. Idx = {0}, count = {1}", idx, count), "idx");
            }
        }

        private BaseTreapNode<T> InsertNode(int idx, BaseTreapNode<T> node, BaseTreapNode<T> tree)
        {
            if(tree == null)
            {
                return node;
            }

            int leftTreeCount = GetTreeLeftCount(tree);
            if(idx - leftTreeCount <= 0)
            {
                tree.Left = InsertNode(idx, node, tree.Left);
                if(tree.Left.Priority < tree.Priority)
                {
                    tree = tree.RotateRight();
                }

                return tree;
            }
            {
                tree.Right = InsertNode(idx - leftTreeCount - 1, node, tree.Right);
                if(tree.Right.Priority < tree.Priority)
                {
                    tree = tree.RotateLeft();
                }
                return tree;
            }
        }

        private void Set(BaseTreapNode<T> node, int idx, T value)
        {
            int leftTreeCount = GetTreeLeftCount(node);

            if (idx == leftTreeCount)
            {
                node.Value = value;    
                node.UpdateAggregate();
                return;
            }

            if (idx < leftTreeCount)
            {
                Set(node.Left, idx, value);
                node.UpdateAggregate();
                return;
            }

            Set(node.Right, idx - (leftTreeCount + 1), value);
            node.UpdateAggregate();
        }

        protected virtual BaseTreapNode<T> Delete(int idx, BaseTreapNode<T> tNode)
		{
            int leftTreeCount = GetTreeLeftCount(tNode);

		    if (idx < leftTreeCount)
		    {
		        tNode.Left = Delete(idx, tNode.Left);
                tNode.UpdateAggregate();
                return tNode;
		    }
		    if (idx > leftTreeCount)
		    {
		        tNode.Right = Delete(idx - leftTreeCount - 1, tNode.Right);
                tNode.UpdateAggregate();
                return tNode;
		    }
		    
		    tNode = tNode.DeleteRoot();

		    return tNode;
		}

        protected int GetTreeLeftCount(BaseTreapNode<T> tNode)
	    {
	        return tNode.Left != null ? tNode.Left.Count : 0;
	    }
	}
}

File: cpp/sa/Treap.cpp
#include <stddef.h>
#include "Treap.h"
#include <stdio.h>

Treap::Treap()
{
  count = 0;
  treapTree = NULL;
}

int Treap::get_tree_left_count(BaseTreapNode* node)
{
  return (node->get_left() != NULL) ? node->get_left()->get_count() : 0;
}

BaseTreapNode* Treap::root()
{
  return treapTree;
}

void Treap::insert(int idx, int data)
{
  insert_node(idx, data);
}

BaseTreapNode* Treap::insert_node(int idx, int data)
{
  BaseTreapNode* node = create_node(data);
  treapTree = insert_node(idx, node, treapTree);
  count++;
  return node;
}

int Treap::get_count()
{
  return  count;
}

int Treap::get_elem(int idx)
{
  BaseTreapNode* treeNode = treapTree;
  
  while (treeNode != NULL)
    {
      int leftTreeCount = get_tree_left_count(treeNode);
      
      if (idx == leftTreeCount)
	{
	  return treeNode->get_value();
	}
      if (idx < leftTreeCount)
	{
	  treeNode = treeNode->get_left();
	}
      else
	{
	  idx -= leftTreeCount + 1;
	  treeNode = treeNode->get_right();
	}
    }
  throw ("Treap key was not found");
}

void Treap::set_elem(int idx, int value)
{
  set_node(treapTree, idx, value);
}

void Treap::DeleteNode(int idx)
{
  treapTree = deleteNode(idx, treapTree);
  count--;
}

BaseTreapNode* Treap::create_node(int data)
{
  return new BaseTreapNode(data);
}

BaseTreapNode* Treap::deleteNode(int idx, BaseTreapNode* node)
{
  int leftTreeCount = this->get_tree_left_count(node);
  
  if (idx < leftTreeCount)
    {
      node->set_left(deleteNode(idx, node->get_left()));
      node->updateAggregate();
      return node;
    }
  if (idx > leftTreeCount)
    {
      node->set_right(deleteNode(idx - leftTreeCount - 1, node->get_right()));
      node->updateAggregate();
      return node;
    }
  
  node = node->deleteRoot();
  
  return node;
}

BaseTreapNode* Treap::insert_node(int idx, BaseTreapNode* node, BaseTreapNode* tree)
{
  if (tree == NULL)
    {
      return node;
    }
  
  int leftTreeCount = get_tree_left_count(tree);
  if (idx - leftTreeCount <= 0)
    {
      tree->set_left(insert_node(idx, node, tree->get_left()));
      if (tree->get_left()->get_priority() < tree->get_priority())
	{
	  tree = tree->rotateRight();
	}
    }
  else
    {
      tree->set_right(insert_node(idx - leftTreeCount - 1, node, tree->get_right()));
      if(tree->get_right()->get_priority() < tree->get_priority())
        {
	  tree = tree->rotateLeft();
        }
    } 
  
  return tree;
}

void Treap::set_node(BaseTreapNode* node, int idx, int value)
{
  int left_tree_count = get_tree_left_count(node);
  if (idx == left_tree_count)
    {
      node->set_value(value);
      node->updateAggregate();
      return; 
    }
  if (idx < left_tree_count)
    {
      set_node(node->get_left(), idx, value);
      node->updateAggregate();
      return ;
    }
  set_node(node->get_right(), idx - (left_tree_count + 1), value);
  node->updateAggregate();
}

Treap::~Treap()
{
  if (treapTree != NULL)
    {
      delete treapTree;
    }
}
